test/ad/utils.jl

```diff
@@ -0,0 +1,163 @@
+    name = f isa Distribution ? nameof(typeof(f)) : nameof(typeof(f(θ...)))
+    return DistSpec(name, f, θ, x, xtrans)
+end
+
+function DistSpec(name::Symbol, f, θ, x, xtrans=nothing)
+    F = f isa Distribution ? typeof(f) : typeof(f(θ...))
+    VF = Distributions.variate_form(F)
+    VS = Distributions.value_support(F)
+    return DistSpec{VF,VS,typeof(f),typeof(θ),typeof(x),typeof(xtrans)}(
+        name, f, θ, x, xtrans
+    )
+end
+
+Distributions.variate_form(::Type{<:DistSpec{VF}}) where VF = VF
+Distributions.value_support(::Type{<:DistSpec{VF,VS}}) where {VF,VS} = VS
+
+# Auxiliary method for vectorizing parameters and samples
+vectorize(v::Number) = [v]
+vectorize(v::Diagonal) = v.diag
+vectorize(v::AbstractVector{<:AbstractMatrix}) = mapreduce(vectorize, vcat, v)
+vectorize(v) = vec(v)
+
+"""
+    unpack(x, inds, original...)
+
+Return a tuple of unpacked parameters and samples in vector `x`.
+
+Here `original` are the original full set of parameters and samples, and
+`inds` contains the indices of the original parameters and samples for which
+a possibly different value is given in `x`. If no value is provided in `x`,
+the original value of the parameter is returned. The values are returned
+in the same order as the original parameters.
+"""
+function unpack(x, inds, original...)
+    offset = 0
+    newvals = ntuple(length(original)) do i
+        if i in inds
+            v, offset = unpack_offset(x, offset, original[i])
+        else
+            v = original[i]
+        end
+        return v
+    end
+    offset == length(x) || throw(ArgumentError())
+
+    return newvals
+end
+
+# Auxiliary methods for unpacking numbers and arrays
+function unpack_offset(x, offset, original::Number)
+    newoffset = offset + 1
+    val = x[newoffset]
+    return val, newoffset
+end
+function unpack_offset(x, offset, original::AbstractArray)
+    newoffset = offset + length(original)
+    val = reshape(x[(offset + 1):newoffset], size(original))
+    return val, newoffset
+end
+function unpack_offset(x, offset, original::AbstractArray{<:AbstractArray})
+    newoffset = offset
+    val = map(original) do orig
+        out, newoffset = unpack_offset(x, newoffset, orig)
+        return out
+    end
+    return val, newoffset
+end
+
+# Run AD tests of a
+function test_ad(dist::DistSpec; kwargs...)
+    @info "Testing: $(dist.name)"
+
+    f = dist.f
+    θ = dist.θ
+    x = dist.x
+    g = dist.xtrans
+
+    # Create function with all possible arguments
+    f_allargs = let f=f, g=g
+        function (x, θ...)
+            dist = f(θ...)
+            xtilde = g === nothing ? x : g(x)
+            if dist isa UnivariateDistribution && xtilde isa AbstractArray
+                result = logpdf.(dist, xtilde)
+            else
+                result = logpdf(dist, xtilde)
+            end
+            return sum(result)
+        end
+    end
+
+    if isempty(θ)
+        # In this case we can only test the gradient with respect to `x`
```

Comment:
We should test wrt to `x` only even if theta is not empty.

Reply:
And we should check if the support is continuous first.

Reply:
The logic is just copied from the previous implementation. There we test wrt `x` if theta is empty and don't check if the support is continuous (see e.g. https://github.com/TuringLang/DistributionsAD.jl/blob/master/test/test_utils.jl#L126 - somehow I forgot to delete the file...). I guess it's OK to not check if the support is continuous, since otherwise we couldn't test any derivatives at all and the `DistSpec` would not be useful. Additionally, both in the old and new implementation we test wrt `x` if theta is not empty and the support is continuous.

Reply:
Even if the old implementation didn't test wrt x only with a non-empty theta, we still should. This is useful for prior distributions in Turing.

Reply:
Hmm I can't follow, why would we want to remove tests? As long as tests pass in a reasonable amount of time (which they do with the new implementation), it should always be better to perform more tests, shouldn't it?

Reply:
Yes I am suggesting performing more tests :) Currently I don't think we test wrt to `x` and only `x` when theta is not empty.

Reply:
Ah I see, somehow I didn't get it and assumed you would want to remove some of the existing tests :laughing: